### ISSUE TITLE
Create separate extended code examples which return more information

### ISSUE THREAD

**Issue by slawa-dev** [bhomnick/identitytw#11]
Some use cases require more than a simple validation check.
e.g. need to know if the ID is from a Taiwanese national or ARC/APRC holder

I suggest the extended code examples to return JSON objects that contain as much info as possible.
Also if validation fails, it should include which check failed (e.g. length, check digit, invalid gender letter, invalid area code)

Encoded in the ID number are:

### Area code
First letter
```
"lettersOrder": "ABCDEFGHJKLMNPQRSTUVXYWZIO",
"areaCodeInfo": [
	{"tw": "臺北市", "en": "Taipei City", "ISO3166-2": "TW-TPE"},         // A
	{"tw": "臺中市", "en": "Taichung City", "ISO3166-2": "TW-TXG"},       // B
	{"tw": "基隆市", "en": "Keelung City", "ISO3166-2": "TW-KEE"},        // C
	{"tw": "臺南市", "en": "Tainan City", "ISO3166-2": "TW-TNN"},         // D
	{"tw": "高雄市", "en": "Kaohsiung City", "ISO3166-2": "TW-KHH"},      // E
	{"tw": "新北市", "en": "New Taipei City", "ISO3166-2": "TW-NWT"},     // F
	{"tw": "宜蘭縣", "en": "Yilan County", "ISO3166-2": "TW-ILA"},        // G
	{"tw": "桃園市", "en": "Taoyuan City", "ISO3166-2": "TW-TAO"},        // H
	{"tw": "新竹縣", "en": "Hsinchu County", "ISO3166-2": "TW-HSQ"},      // J
	{"tw": "苗栗縣", "en": "Miaoli County", "ISO3166-2": "TW-MIA"},       // K
	{"tw": "臺中縣", "en": "Taichung County", "ISO3166-2": "TW-TXQ"},     // L (no longer issued)
	{"tw": "南投縣", "en": "Nantou County", "ISO3166-2": "TW-NAN"},       // M
	{"tw": "彰化縣", "en": "Changhua County", "ISO3166-2": "TW-CHA"},     // N
	{"tw": "雲林縣", "en": "Yunlin County", "ISO3166-2": "TW-YUN"},       // P
	{"tw": "嘉義縣", "en": "Chiayi County", "ISO3166-2": "TW-CYQ"},       // Q
	{"tw": "臺南縣", "en": "Tainan County", "ISO3166-2": "TW-TNN"},       // R (no longer issued)
	{"tw": "高雄縣", "en": "Kaohsiung County", "ISO3166-2": "TW-KHQ"},    // S (no longer issued)
	{"tw": "屏東縣", "en": "Pingtung County", "ISO3166-2": "TW-PIF"},     // T
	{"tw": "花蓮縣", "en": "Hualien County", "ISO3166-2": "TW-HUA"},      // U
	{"tw": "臺東縣", "en": "Taitung County", "ISO3166-2": "TW-TTT"},      // V
	{"tw": "澎湖縣", "en": "Penghu County", "ISO3166-2": "TW-PEN"},       // X
	{"tw": "陽明山管理局", "en": "Yangmingshan Management Bureau", "ISO3166-2": ""}, // Y (no longer issued, no ISO 3166-2 code)
	{"tw": "金門縣", "en": "Kinmen County", "ISO3166-2": "TW-KIN"},       // W
	{"tw": "連江縣", "en": "Lienchiang County", "ISO3166-2": "TW-LIE"},   // Z
	{"tw": "嘉義市", "en": "Chiayi City", "ISO3166-2": "TW-CYI"},         // I
	{"tw": "新竹市", "en": "Hsinchu City", "ISO3166-2": "TW-HSZ"},        // O
]
```

### Gender & Nationality
Second letter (and third letter in new resident certificate format)

#### Taiwanese national with household registration
1 = male
2 = female

#### National without household registration (NWOHR) 
[OLD FORMAT]
A = male
B = female

[NEW FORMAT]
8 = male
9 = female
... followed by 7 for NWOHR, 8 for HK+MO, 9 for CN

#### Alien Resident (ARC/APRC)
[OLD FORMAT]
C = male
D = female

[NEW FORMAT]
8 = male
9 = female
... followed by 0 to 6 [encoding of 0-6 unknown yet]

**Comment by slawa-dev** [comment]
I wrote an API example for CloudFlare Workers:

https://twid.slawa.workers.dev/IDNUMBER

I can share the code after I clean it up a bit.

**Comment by bhomnick** [comment]
:+1:

**Comment by bhomnick** [comment]
> e.g. need to know if the ID is from a Taiwanese national or ARC/APRC holder

I think this case should explicitly be avoided.  The point of this library is to make no distinction between ARC and national ID holders.

**Comment by slawa-dev** [comment]
I also would like that online services, etc. would not make that distinction.

Unfortunately some services require additional information from ARC holders due to laws & regulations. I worry those will avoid this library and might use something that does not validate all ARC cases properly.
This is why I want the extended code to be optional for those who need it.

The short and simple code should be promoted by default.

**Comment by slawa-dev** [comment]
> I can share the code after I clean it up a bit.

https://gist.github.com/slawa-dev/d0b4143baf7429584482efcb145a2d48

**Comment by fifieldt** [comment]
TL;DR I'm with Ben on this one.

> Unfortunately some services require additional information from ARC holders due to laws & regulations. I worry those will avoid this library and might use something that does not validate all ARC cases properly.

I've read a lot of javascript for ID validation in forms over recent years. My impression is that people making forms involving ID just copy and paste code directly in from wherever they find it. I don't think I saw one that used an external validation library for the ID logic (some use generic validation libraries, but the ID validation logic was always in the site's own bundle). So, I don't think there's a problem about use -- we just want people to do what they do right now: copy and paste the nicest available code without thinking too much. Only, this time they'll be copying code in that does not discriminate between the various types of legal resident.

There are relatively few online services that need to make a distinction between residents and citizens due to a law or regulation (I guess many that do so right now have no legal basis to do so). It is a corner case that needs to be satisfied, and I personally appreciate and use the tool on slawa.dev, but as I understand it not the original intention of this particular site.

**Comment by slawa-dev** [comment]
Good points.
I will update the post on my site to include the new format and link to the website of this project.
